README.md
# ABC Manager - Local Partner - WordPress Plugin

[![Author](http://img.shields.io/badge/author-@angrybytes-blue.svg?style=flat-square)](https://twitter.com/angrybytes)
[![Software License](https://img.shields.io/badge/license-proprietary-brightgreen.svg?style=flat-square)](LICENSE.md)

WordPress Plugin to post new updates to the ABC Manager of NH/AT5

## Setup

> This plugin requires you to have the following minimal requirements
> > PHP: **7.3** | WordPress: **5.7**

### Download and installation

You can download this plugin via the `Code` button and then select `download zip`

![Clone or Download](./docs/images/clone-download.png)

To install the plugin into your WordPress environment you go to `Plugins -> Add New -> Upload Plugin`. Here you can
select the downloaded ZIP file and click `Install Now`

## Activate Plugin

To activate the plugin you can go to the plugins overview and select `Activate`
![Activate plugin](./docs/images/plugins.png)

## Plugin settings

![Plugin](./docs/images/plugin.png)

You can find the settings of the plugin in `Settings -> ABC Manager`. The values of the settings you have to fill in
have been given to you by your contact person from NH Nieuws. If you entered the correct credentials the  ABC Manager
status should be `UP` and the Credentials should be `Valid`

---

## How it works

Every time you publish and new article or update a published article this article will be send to ABC Manager.
Upon receipt, the article will be reviewed bij the editors and they will be able to publish the article.
Image/Video/Audio files will be uploaded in their own system so you don't have to worry about increased network traffic.

If NH Nieuws publishes an article about your region, this article will be automatically send to your WordPress environment.
Dependable on the settings in ABC Manager the article will get the status `Published` or `Draft`

### Errors

Do you have problems installing the plugin or experience errors in the syncing of the articles, please contact
[NH Nieuws](https://www.nhnieuws.nl/contact).

&copy;  NH Media

## Dutch documentation

Voor de Nederlandse
documentatie, [klikt u hier](https://github.com/rtvnh/abc-local-wp-custom/blob/develop/docs/nl.md).


CHANGELOG.md
# Changelog

## 0.1.13
* Send articles that are planned for the future to ABC manager

## 0.1.12
* Add author name to the post content

## 0.1.9
* Add custom video embed code to the post content

## 0.1.7
* Use correct slug for audio description

## 0.1.5
* Use different hook to include the post meta

## 0.1.4
* Add image caption
* Audio-Items update

## 0.1.2
* Prevent same article posting multiple times

## 0.1.0
* Added code to add the audio-item post type to the corresponding article when posting to abc-manager


class-abclocalpartnerwp-updater.php
<?php
/**
 * The plugin updater. file
 *
 * This file is read by WordPress to automatic update this plugin with the latest version available on GitHub.
 *
 * @package Plugin_ABC_Manager_Local_Partner
 * @since   0.4.0
 */

/**
 * Updater class to deploy the WordPress ABC Local Partner plugin using GitHub.
 */
class AbcLocalPartnerWp_Updater {

	/**
	 * The plugin's file path.
	 *
	 * @var string
	 */
	private $file;

	/**
	 * The plugin's metadata.
	 *
	 * Parsed from the starting comment in the plugin file.
	 *
	 * @var string[]
	 */
	private $plugin;

	/**
	 * The plugin's name.
	 *
	 * Parsed from the plugin file path.
	 *
	 * @var string
	 */
	private $basename;

	/**
	 * Flag that indicates if the plugin is currently active.
	 *
	 * @var bool
	 */
	private $active;

	/**
	 * The GitHub username.
	 *
	 * @var string
	 */
	private $username;

	/**
	 * The GitHub repository.
	 *
	 * @var string
	 */
	private $repository;

	/**
	 * The latest plugin version retrieved from GitHub.
	 *
	 * @var string[]|null
	 */
	private $latest_version;

	/**
	 * Initiate plugin properties.
	 *
	 * @param string $file The plugin's file path.
	 */
	public function __construct( string $file ) {
		$this->file = $file;

		// Postpone the final plugin properties until "admin_init" is fired. This ensures that "get_plugin_data" is available.
		add_action( 'admin_init', array( $this, 'set_plugin_properties' ) );
	}

	/**
	 * Finalize plugin properties.
	 */
	public function set_plugin_properties(): void {
		$this->plugin   = get_plugin_data( $this->file );
		$this->basename = plugin_basename( $this->file );
		$this->active   = is_plugin_active( $this->basename );
	}

	/**
	 * Set the GitHub username.
	 *
	 * @param string $username The GitHub username.
	 */
	public function set_username( string $username ): void {
		$this->username = $username;
	}

	/**
	 * Set the GitHub repository.
	 *
	 * @param string $repository THe GitHub repository.
	 */
	public function set_repository( string $repository ): void {
		$this->repository = $repository;
	}

	/**
	 * Fetch the latest plugin version from GitHub.
	 *
	 * @return string[]
	 */
	private function get_latest_plugin_version(): array {
		if ( is_null( $this->latest_version ) ) {
			$request_uri = sprintf(
				'https://api.github.com/repos/%s/%s/releases',
				$this->username,
				$this->repository
			);

			// Fetch the plugin releases from GitHub.
			$response = json_decode(
				wp_remote_retrieve_body( wp_remote_get( $request_uri ) ),
				true
			);

			// If we have a valid response.
			if ( is_array( $response ) ) {
				// Store the latest release.
				$this->latest_version = current( $response );
			}
		}

		return $this->latest_version;
	}

	/**
	 * Initialize the plugin updater.
	 */
	public function initialize(): void {
		add_filter( 'pre_set_site_transient_update_plugins', array( $this, 'modify_transient' ), 10, 1 );
		add_filter( 'plugins_api', array( $this, 'plugin_popup' ), 10, 3 );
		add_filter( 'upgrader_post_install', array( $this, 'after_install' ), 10, 3 );
	}

	/**
	 * Modify a transient with our plugin information.
	 *
	 * @param stdClass $transient The transient object.
	 *
	 * @return mixed
	 */
	public function modify_transient( stdClass $transient ) {
		// Did WordPress check for updates?
		if ( property_exists( $transient, 'checked' ) && $transient->checked ) {
			// Fetch the latest plugin version.
			$latest_version = $this->get_latest_plugin_version();

			// Compare with our current version.
			$out_of_date = version_compare(
				$latest_version['tag_name'],
				$transient->checked[ $this->basename ],
				'gt'
			);

			if ( $out_of_date ) {
				// Our plugin is out of date, setup our plugin information.
				$new_files = $latest_version['zipball_url'];
				$slug      = current( explode( '/', $this->basename ) );

				$plugin = array(
					'url'         => $this->plugin['PluginURI'],
					'slug'        => $slug,
					'package'     => $new_files,
					'new_version' => $latest_version['tag_name'],
				);

				// Modify the transient with our updated plugin information.
				$transient->response[ $this->basename ] = (object) $plugin;
			}
		}

		return $transient;
	}

	/**
	 * Provide WordPress plugin popup with information about our plugin.
	 *
	 * @param false|stdClass|mixed[] $result    The result object or array.
	 * @param string                 $action        The type of information being requested from the Plugin Installation API.
	 * @param object                 $args          Plugin API arguments.
	 *
	 * @return false|mixed|object
	 */
	public function plugin_popup( $result, string $action, object $args ) {
		if ( 'plugin_information' !== $action ) {
			return false;
		}

		// Check if the slug matches our plugin.
		if ( ! empty( $args->slug ) && current( explode( '/', $this->basename ) ) === $args->slug ) {
			$latest_version = $this->get_latest_plugin_version();

			$plugin = array(
				'name'              => $this->plugin['Name'],
				'slug'              => $this->basename,
				'requires'          => '5.3',
				'tested'            => '5.4',
				'version'           => $latest_version['tag_name'],
				'author'            => $this->plugin['AuthorName'],
				'author_profile'    => $this->plugin['AuthorURI'],
				'last_updated'      => $latest_version['published_at'],
				'homepage'          => $this->plugin['PluginURI'],
				'short_description' => $this->plugin['Description'],
				'sections'          => array(
					'Description' => $this->plugin['Description'],
					'Updates'     => $latest_version['body'],
				),
				'download_link'     => $latest_version['zipball_url'],
			);

			// Return our plugin formation.
			return (object) $plugin;
		}

		// Otherwise, return the default information.
		return $result;
	}

	/**
	 * Move and reactive our plugin after installation.
	 *
	 * @param bool     $response       The installation response.
	 * @param string[] $hook_extra     Extra arguments passed to hooked filters.
	 * @param mixed[]  $result         Installation result data.
	 *
	 * @return mixed[]
	 */
	public function after_install( bool $response, array $hook_extra, array $result ): array {
		global $wp_filesystem;

		// Move our updated plugin files to the plugin directory.
		$install_directory = plugin_dir_path( $this->file );
		$wp_filesystem->move( $result['destination'], $install_directory );
		$result['destination'] = $install_directory;

		// If our plugin was active.
		if ( $this->active ) {
			// Reactive it.
			activate_plugin( $this->basename );
		}

		return $result;
	}
}


abc-local-partner-wp-plugin.php
<?php
/**
 * The plugin bootstrap file
 *
 * This file is read by WordPress to generate the plugin information in the plugin
 * admin area. This file also includes all the dependencies used by the plugin,
 * registers the activation and deactivation functions, and defines a function
 * that starts the plugin.
 *
 * @link              https://github.com/rtvnh/abc-local-wp-custom
 * @since             0.1.0
 * @package           Plugin_ABC_Manager_Local_Partner
 *
 * @wordpress-plugin
 * Plugin Name:         ABC Manager - Custom Local Partner
 * Plugin URI:          https://github.com/rtvnh/abc-local-wp-custom
 * Description:         WordPress Plugin to post new updates to the ABC Manager of NH/AT5
 * Version:             0.1.13
 * Author:              AngryBytes B.V.
 * Author URI:          https://angrybytes.com
 * License:             GPL-2.0+
 * Text Domain:         abclocalpartner
 * Requires at least:   5.7
 * Requires PHP:        7.3
 */

// Configure our plugin updater.
require_once plugin_dir_path( __FILE__ ) . '/class-abclocalpartnerwp-updater.php';

$updater = new AbcLocalPartnerWp_Updater( __FILE__ );
$updater->set_username( 'rtvnh' );
$updater->set_repository( 'abc-local-wp-custom' );
$updater->initialize();

$abc_post_status = false;

/**
 * Register our plugin settings.
 */
function abclocalpartner_register_settings(): void {
	add_option( 'abclocalpartner_option_abc_url' );
	add_option( 'abclocalpartner_option_partner_name' );
	add_option( 'abclocalpartner_option_partner_client_id' );
	add_option( 'abclocalpartner_option_partner_client_secret' );
	add_option( 'abclocalpartner_option_access_token' );
	add_option( 'abclocalpartner_option_region_name' );

	register_setting(
		'abclocalpartner_options_group',
		'abclocalpartner_option_abc_url'
	);
	register_setting(
		'abclocalpartner_options_group',
		'abclocalpartner_option_partner_name'
	);
	register_setting(
		'abclocalpartner_options_group',
		'abclocalpartner_option_partner_client_id'
	);
	register_setting(
		'abclocalpartner_options_group',
		'abclocalpartner_option_partner_client_secret'
	);
	register_setting(
		'abclocalpartner_options_group',
		'abclocalpartner_option_access_token'
	);
	register_setting(
		'abclocalpartner_options_group',
		'abclocalpartner_option_region_name'
	);
}

add_action( 'admin_init', 'abclocalpartner_register_settings' );

/**
 * Register the options page for our plugin.
 */
function abclocalpartner_register_options_page(): void {
	add_options_page( 'ABC Manager - Settings', 'ABC Manager', 'manage_options', 'abclocalpartner', 'abclocalpartner_options_page' );
}

add_action( 'admin_menu', 'abclocalpartner_register_options_page' );

/**
 * Generate the options page for our plugin.
 */
function abclocalpartner_options_page(): void {
	?>
	<div>
		<h1>RTV NH/AT5 - ABC Manager</h1>
		<div>
			<img src="<?php echo esc_url( plugin_dir_url( __FILE__ ) ); ?>/assets/images/nh-at5-logo.png" height="64" width="64" alt="NH"/>
		</div>
		<form method="post" action="options.php">
			<?php settings_fields( 'abclocalpartner_options_group' ); ?>
			<h2>ABC Manager Options</h2>
			<p>Please adjust the settings for a connection with ABC.</p>
			<table class="form-table" role="presentation">
				<tbody>
					<tr>
						<th scope="row">
							<label for="abclocalpartner_option_abc_url">API URL</label>
						</th>
						<td>
							<input type="text" id="abclocalpartner_option_abc_url" name="abclocalpartner_option_abc_url"
								class="regular-text"
								value="<?php echo esc_url( get_option( 'abclocalpartner_option_abc_url' ) ); ?>"/>
						</td>
					</tr>
					<tr>
						<th scope="row">
							<label for="abclocalpartner_option_partner_name">Partner name</label>
						</th>
						<td>
							<input type="text" id="abclocalpartner_option_partner_name"
								name="abclocalpartner_option_partner_name" class="regular-text"
								value="<?php echo esc_attr( get_option( 'abclocalpartner_option_partner_name' ) ); ?>"/>
						</td>
					</tr>
					<tr>
						<th scope="row">
							<label for="abclocalpartner_option_partner_client_id">Client ID</label>
						</th>
						<td>
							<input type="text" id="abclocalpartner_option_partner_client_id"
								name="abclocalpartner_option_partner_client_id" class="regular-text"
								value="<?php echo esc_attr( get_option( 'abclocalpartner_option_partner_client_id' ) ); ?>"/>
						</td>
					</tr>
					<tr>
						<th scope="row">
							<label for="abclocalpartner_option_partner_client_secret">Client Secret</label>
						</th>
						<td>
							<input type="password" id="abclocalpartner_option_partner_client_secret"
								name="abclocalpartner_option_partner_client_secret" class="regular-text"
								value="<?php echo esc_attr( get_option( 'abclocalpartner_option_partner_client_secret' ) ); ?>"/>
						</td>
					</tr>
					<tr>
						<th scope="row">
							<label for="abclocalpartner_option_region_name">Region taxonomy name</label>
						</th>
						<td>
							<input type="text" id="abclocalpartner_option_region_name"
								name="abclocalpartner_option_region_name" class="regular-text"
								value="<?php echo esc_attr( get_option( 'abclocalpartner_option_region_name' ) ); ?>"/>
						</td>
					</tr>
					<style type="text/css">
						.status-dot {
							height: 10px;
							width: 10px;
							border-radius: 1000px;
							display: inline-block;
						}

						.status-dot.is-red {
							background-color: red;
						}

						.status-dot.is-green {
							background-color: green;
						}
					</style>
					<?php
					// Are ABC Manager and WordPress connected?
					$online       = check_abc_status();
					$status       = $online ? 'Up' : 'Down';
					$status_color = $online ? 'green' : 'red';
					?>
					<tr>
						<th>
							ABC Manager status:
						</th>
						<td>
							<span class="status-dot is-<?php echo esc_attr( $status_color ); ?>"></span> <?php echo esc_html( $status ); ?>
						</td>
					</tr>

					<?php
					// Are ABC Manager credentials valid?
					$valid        = check_credentials();
					$status       = $valid ? 'Valid' : 'Invalid';
					$status_color = $valid ? 'green' : 'red';
					?>
					<tr>
						<th>
							Credentials:
						</th>
						<td>
							<span class="status-dot is-<?php echo esc_attr( $status_color ); ?>"></span> <?php echo esc_html( $status ); ?>
						</td>
					</tr>
				</tbody>
			</table>
			<?php submit_button(); ?>
		</form>
	</div>
	<?php
}

/**
 * Get a bearer token from ABC Manager.
 *
 * @param bool $force_new  With this option set to true, you can force to always retrieve a new bearer token.
 *
 * @return string|null
 */
function get_abc_bearer_token( $force_new = false ) {
	$bearer_token = get_option( 'abclocalpartner_option_access_token' );

	if ( ! empty( $bearer_token ) && ! $force_new ) {
		return $bearer_token;
	}

	$api_endpoint  = get_option( 'abclocalpartner_option_abc_url' );
	$client_id     = get_option( 'abclocalpartner_option_partner_client_id' );
	$client_secret = get_option( 'abclocalpartner_option_partner_client_secret' );

	$raw_response = wp_remote_post(
		$api_endpoint . '/oauth2/token',
		array(
			'headers' => array_merge( get_environment_headers(), array( 'Content-Type' => 'application/x-www-form-urlencoded' ) ),
			'body'    => array(
				'grant_type'    => 'client_credentials',
				'client_id'     => $client_id,
				'client_secret' => $client_secret,
				'scope'         => 'partners',
			),
		)
	);

	if ( is_array( $raw_response ) ) {
		try {
			$data = json_decode( $raw_response['body'], true, 512, JSON_THROW_ON_ERROR );

			if ( array_key_exists( 'access_token', $data ) ) {
				update_option( 'abclocalpartner_option_access_token', $data['access_token'] );

				return $data['access_token'];
			}
		} catch ( Exception $e ) {
			return null;
		}
	}

	return null;
}

/**
 * This does a simple check if the credentials are valid.
 *
 * @param int $attempts Total of attempts.
 *
 * @return bool
 */
function check_credentials( $attempts = 0 ): bool {
	$token = get_abc_bearer_token();

	$api_endpoint = get_option( 'abclocalpartner_option_abc_url' );
	$partner_name = get_option( 'abclocalpartner_option_partner_name' );

	try {
		$response = wp_remote_get(
			sprintf( '%s/partner/article?partner=%s', $api_endpoint, $partner_name ),
			array(
				'headers' => array_merge( get_environment_headers(), array( 'Authorization' => sprintf( 'Bearer %s', $token ) ) ),
			)
		);

		if ( $response instanceof WP_Error ) {
			return false;
		}

		if ( array_key_exists( 'response', $response ) ) {
			switch ( $response['response']['code'] ) {
				case 204:
					return true;
				case 401:
					if ( 1 === $attempts ) {
						return false;
					}

					get_abc_bearer_token( true );

					return check_credentials( 1 );
				default:
					return false;
			}
		}
	} catch ( \Exception $e ) {
		return false;
	}

	return false;
}

/**
 * Used for development purposes only, adds Host header to all requests
 *
 * @return array|string[]
 */
function get_environment_headers(): array {
	return wp_get_environment_type() === 'local' ? array( 'Host' => 'partner.test' ) : array();
}

/**
 * This does a simple check if ABC Manager is available
 *
 * @return bool
 */
function check_abc_status(): bool {
	$api_endpoint = get_option( 'abclocalpartner_option_abc_url' );
	$partner_name = get_option( 'abclocalpartner_option_partner_name' );

	try {
		$response = wp_remote_get(
			sprintf( '%s/partner/article?partner=%s', $api_endpoint, $partner_name ),
			array(
				'headers' => get_environment_headers(),
			)
		);

		if ( $response instanceof WP_Error ) {
			return false;
		}

		if ( array_key_exists( 'response', $response ) ) {
			switch ( $response['response']['code'] ) {
				case 401:
					return true;
				default:
					return false;
			}
		}
	} catch ( \Exception $e ) {
		return false;
	}

	return false;
}

/**
 * Post an article to ABC Manager.
 *
 * @param WP_Post  $post             The WordPress post instance.
 * @param string[] $post_galleries   A list of galleries from the post content.
 * @param string   $post_featured    An URL for the featured image in the post.
 * @param int      $attempts         The amount of retry attempts.
 *
 * @return bool
 */
function post_article_to_abc_manager(
    WP_Post $post,
    array $post_galleries,
    string $post_featured,
    int $attempts = 0
): bool {
	$galleries_json  = wp_json_encode( $post_galleries );
	$api_endpoint    = get_option( 'abclocalpartner_option_abc_url' );
	$partner_name    = get_option( 'abclocalpartner_option_partner_name' );
	$region_taxonomy = get_option( 'abclocalpartner_option_region_name' );
	$bearer_token    = get_abc_bearer_token();
	$regions         = get_the_terms( $post, $region_taxonomy );

	if ( empty( $bearer_token ) ) {
		return false;
	}

    global $wpdb;

    $postMeta = get_post_meta($post->ID);

    // Prevent duplicate posts to abc for new articles
    if (!isset($postMeta['_edit_last'])) {
        return false;
    }

    $videoEmbedCodes = $wpdb->get_results("
        SELECT meta_value
        FROM wp_postmeta
        WHERE wp_postmeta.post_id = {$post->ID}
        AND wp_postmeta.meta_key = 'wpcf-video-embedcode'
    ");

    if (count($videoEmbedCodes) > 0) {
        foreach ($videoEmbedCodes as $videoEmbedCode) {
            // Add the content as custom html.
            $post->post_content .= "<!-- wp:custom-html -->
                $videoEmbedCode->meta_value
                <!-- /wp:custom-html -->
            ";
        }
    }

    $podcastMeta = $wpdb->get_results("
        SELECT post_id, meta_key, meta_value
        FROM wp_postmeta
        INNER JOIN wp_toolset_connected_elements
        ON wp_postmeta.post_id = wp_toolset_connected_elements.element_id
        INNER JOIN wp_toolset_associations
        ON wp_toolset_connected_elements.group_id = wp_toolset_associations.child_id
        WHERE wp_toolset_associations.parent_id = (
            SELECT group_id
            FROM wp_toolset_connected_elements
            WHERE wp_toolset_connected_elements.element_id = {$post->ID}
        )
        AND wp_toolset_associations.relationship_id = (
            SELECT id
            FROM wp_toolset_relationships
            WHERE wp_toolset_relationships.slug = 'audio-item'
        )
    ");

    if (count($podcastMeta) > 0) {
        $podcastData = [];

        foreach ($podcastMeta as $meta) {
            if (!empty(trim($meta->meta_value))) {
                $podcastData[$meta->post_id][$meta->meta_key] = $meta->meta_value;
            }
        }

        foreach ($podcastData as $data) {
            $date = wp_date('d F Y', $data['wpcf-datum-uitzending-fragment']);

            $post->post_content .= "<!-- wp:custom-html -->
            <custom-html>
              <h1>Uitzending - {$date}</h1>
              <p>{$data['wpcf-beschrijving-audio']}</p>
            </custom-html>
            <!-- /wp:custom-html -->
            <!-- wp:audio -->
            <figure class='wp-block-audio'><audio controls src='{$data['wpcf-audiobestand']}'></audio></figure>
            <!-- /wp:audio -->
            ";
        }
    }

    // Add author name to post content
    $post->author_name = get_the_author_meta('display_name', $post->post_author);

	$response = wp_remote_post(
		$api_endpoint . '/partner/article',
		array(
			'body'    => array(
				'partner'   => $partner_name,
				'content'   => wp_json_encode( $post ),
				'featured'  => $post_featured,
				'galleries' => $galleries_json,
				'regions'   => wp_json_encode( $regions ),
			),
			'headers' => array_merge( get_environment_headers(), array( 'Authorization' => 'Bearer ' . $bearer_token ) ),
		)
	);

	if ( $response instanceof WP_Error ) {
		return false;
	}

	if ( 'Partner not authorized.' === $response['body'] ) {
		return true;
	}
	if ( 'Unauthorized' === $response['body'] ) {
		if ( 0 === $attempts ) {
			get_abc_bearer_token( true );

			post_article_to_abc_manager( $post, $post_galleries, $post_featured, $attempts + 1 );
		}

		return false;
	}

	return true;
}

/**
 * Register a hook on "save_post".
 *
 * @param WP_Post $post The WordPress post.
 */
function abclocalpartner_post_to_abc( WP_Post $post ): void {
	if ( ! empty( get_option( 'abclocalpartner_option_abc_url' ) ) &&
		! empty( get_option( 'abclocalpartner_option_partner_client_id' ) ) &&
		! empty( get_option( 'abclocalpartner_option_partner_client_secret' ) )
	) {
        $postStatus = get_post_status( $post );

		if ( ($postStatus === 'publish' || $postStatus === 'future') && $post->post_type === 'post' ) {
			$post_galleries = get_post_galleries( $post );
			$post_featured  = get_the_post_thumbnail_url( $post );

			if ( is_bool( $post_featured ) ) {
				$post_featured = '';
			}

			global $abc_post_status;
			$abc_post_status = post_article_to_abc_manager(
				$post,
				$post_galleries,
				$post_featured
			);
		}
	}
}

/**
 * Triggers only on gutenberg save calls
 *
 * @param   WP_Post $post WordPress post.
 */
function gutenberg_post_to_abc( int $postId, WP_Post $post, bool $update ): void {
	// Prevent save calls from ABC Manager to be also send to ABC Manager back again.
    // phpcs:ignore
	if ( defined( 'REST_REQUEST' ) && REST_REQUEST && isset( $_GET['abc'] ) ) {
		return;
	}

	abclocalpartner_post_to_abc( $post );
}

add_action( 'wp_insert_post', 'gutenberg_post_to_abc', 10, 3 );

/**
 * Allow iframe HTML tags.
 *
 * @param string[] $tags    A list of HTML tags.
 * @param string   $context The context to judge allowed tags by.
 *
 * @return mixed[]
 */
function prefix_add_source_tag( array $tags, string $context ): array {
	if ( 'post' === $context ) {
		$tags['iframe'] = array(
			'src'    => true,
			'srcdoc' => true,
			'width'  => true,
			'height' => true,
		);
	}
	return $tags;
}

add_filter( 'wp_kses_allowed_html', 'prefix_add_source_tag', 10, 2 );

/**
 * Add custom meta values for posts from ABC manager
 *
 * @param $response
 * @param $object
 * @param $request
 * @return mixed
 */
function default_meta_from_abc_post( $response, $object, $request) {
    if (
        isset($response[ 'type' ]) &&
        $response[ 'type' ] === 'post' &&
        defined( 'REST_REQUEST' ) &&
        REST_REQUEST &&
        isset( $_GET[ 'abc' ] )
    ) {
        $post_id = $response[ 'id' ];
        // Add citation to featured image
        if (isset($_GET[ 'imageCitation' ])) {
            update_post_meta($post_id, 'wpcf-bronvermelding-foto', $_GET[ 'imageCitation' ]);
        }

        // New post, set wpcf-laat-artikel-niet-zien-op-voorpagina to false
        if ($request->get_route() === '/wp/v2/posts') {
            update_post_meta($post_id, 'wpcf-laat-artikel-zien-op-voorpagina', 0);
        }
    }

    return $response;
}

add_filter( 'rest_pre_echo_response', 'default_meta_from_abc_post', 10, 3 );
